feat: add Meilisearch-compatible search engine

regisb opened the pull request:
## Description

The goal of this change is to introduce a search engine that is compatible with the edx-search API but that uses Meilisearch instead of Elasticsearch. That way, we can replace one by the other across edx-platform by simply changing a single SEARCH_ENGINE django setting.

There are a couple of differences between Meilisearch and Elasticsearch:

1. Filterable attributes must be defined explicitly.
2. No support for datetime objects, which must be converted to timestamps (with an extra field to store the timezone).
3. No special characters allowed in the primary key values, such that we must hash course IDs before we can use them as primary key values.

Note that this PR does not introduce any breaking change. This is an opt-in engine that anyone is free to use. There is some setup work for every search feature: see the engine module documentation for more information.


## Supporting information

This change was discussed here: https://github.com/openedx/frontend-app-authoring/issues/1334#issuecomment-2401805382

## Testing instructions

Set the following setting:

    SEARCH_ENGINE = "openedx.core.djangoapps.content.search.engine.MeilisearchEngine"

Then, follow the instructions from the engine module to try out the different search features.

## Deadline

I'd like to merge this in time for the Sumac branch (October 23rd). If it's not possible, then I might opt to install this feature in Tutor via a separate Python package instead.

## Open questions

As far as I'm concerned there are a couple of open questions for this PR:

- [ ] Is edx-platform the right place to add this module? (I think it is, because edx-search is used only in edx-platform). If yes, did I pick the right location in the edx-platform file hierarchy, and did I pick the right module name?
- [ ] Are there other features which make use of Elasticsearch and for which we should test this Meilisearch engine? I am aware of the following:
    - [ ] student notes
    - [ ] forum

regisb committed:
feat: add Meilisearch-compatible search engine

The goal of this change is to introduce a search engine that is
compatible with the edx-search API but that uses Meilisearch instead of
Elasticsearch. That way, we can replace one by the other across
edx-platform by simply changing a single SEARCH_ENGINE django setting.

There are a couple of differences between Meilisearch and Elasticsearch:

1. Filterable attributes must be defined explicitly.
2. No support for datetime objects, which must be converted to
   timestamps (with an extra field to store the timezone).
3. No special characters allowed in the primary key values, such that we
   must hash course IDs before we can use them as primary key values.

Note that this PR does not introduce any breaking change. This is an
opt-in engine that anyone is free to use. There is some setup work for
every search feature: see the engine module documentation for more
information.

See the corresponding conversation here:
https://github.com/openedx/frontend-app-authoring/issues/1334#issuecomment-2401805382

openedx-webhooks commented:
Thanks for the pull request, @regisb!

### What's next?

*Please work through the following steps to get your changes ready for engineering review:*

#### :radio_button: Get product approval

If you haven't already, [check this list](https://openedx.atlassian.net/wiki/spaces/COMM/pages/3875962884/How+to+submit+an+open+source+contribution+for+Product+Review#Does-my-contribution-require-Product-Review%3F) to see if your contribution needs to go through the product review process.

- If it does, you'll need to submit a product proposal for your contribution, and have it reviewed by the [Product Working Group](https://openedx.atlassian.net/wiki/spaces/COMM/pages/3449028609/Product+Working+Group).
    - This process (including the steps you'll need to take) is documented [here](https://openedx.atlassian.net/wiki/spaces/COMM/pages/3875962884/How+to+submit+an+open+source+contribution+for+Product+Review#Product-Review-Process).
- If it doesn't, simply proceed with the next step.

#### :radio_button: Provide context

To help your reviewers and other members of the community understand the purpose and larger context of your changes, feel free to add as much of the following information to the PR description as you can:

- Dependencies
  > This PR must be merged before / after / at the same time as ...
- Blockers
  > This PR is waiting for OEP-1234 to be accepted.
- Timeline information
  > This PR must be merged by XX date **because** ...
- Partner information
  > This is for a course on edx.org.
- Supporting documentation
- Relevant [Open edX discussion forum](https://discuss.openedx.org/) threads





#### :radio_button: Get a green build

If one or more checks are failing, continue working on your changes until this is no longer the case and your build turns green.



#### :radio_button: Let us know that your PR is ready for review:

### Who will review my changes?


This repository is currently maintained by `@openedx/wg-maintenance-edx-platform`.  Tag them in a comment and let them know that your changes are ready for review.


### Where can I find more information?

If you'd like to get more details on all aspects of the review process for open source pull requests (OSPRs), check out the following resources:

- [Overview of Review Process for Community Contributions](https://docs.openedx.org/en/latest/developers/references/developer_guide/process/FAQ-about-pull-requests.html)
- [Pull Request Status Guide](https://docs.openedx.org/en/latest/developers/references/developer_guide/process/pull-request-statuses.html)
- [Making changes to your pull request](https://docs.openedx.org/en/latest/documentors/how-tos/make_changes_to_your_pull_request.html)

### When can I expect my changes to be merged?

Our goal is to get community contributions seen and reviewed as efficiently as possible.

However, the amount of time that it takes to review and merge a PR can vary significantly based on factors such as:

- The size and impact of the changes that it introduces
- The need for product review
- Maintenance status of the parent repository

:bulb: *As a result it may take up to several weeks or months to complete a review and merge your PR.*

<!-- comment:external_pr -->
<!-- data: eyJkcmFmdCI6IGZhbHNlfQ== -->

pomegranited commented on openedx/core/djangoapps/content/search/engine.py:
```diff
@@ -0,0 +1,484 @@
+switch to this engine, you should run a Meilisearch instance and define the following
+setting:
+
+    SEARCH_ENGINE = "openedx.core.djangoapps.content.search.engine.MeilisearchEngine"
+
+You will then need to create the new indices by running:
+
+    ./manage.py lms shell -c "from openedx.core.djangoapps.content.search import engine; engine.create_indexes()"
```
To use Meilisearch in the LMS, we also need to move the [meilisearch settings in the CMS env](https://github.com/openedx/edx-platform/blob/af69cbe6714f21d70d77a27ad75d00bdfee3911d/cms/envs/common.py#L2920-L2933) to `lms.env.common`.

regisb replied:
Good point. But then we would have to have the same settings in cms.envs.common, right? Right above the line `from lms.envs.common import (...)` there is a big warning:

> Although this module itself may not use these imported variables, other dependent modules may. 
> Warning: Do NOT add any new variables to this list. This is incompatible with future plans to
> have more logical separation between LMS and Studio (CMS). It is also incompatible with the
> direction documented in OEP-45: Configuring and Operating Open edX:
> https://open-edx-proposals.readthedocs.io/en/latest/oep-0045-arch-ops-and-config.html

OEP-45 says we should use a defaults.py settings file: https://open-edx-proposals.readthedocs.io/en/latest/architectural-decisions/oep-0045-arch-ops-and-config.html#configuration

But there is no such settings file yet. I'd rather not use this PR to introduce this new module (scope creep...). What I can do is move all those settings to content/search/api.py, like this:

```
MEILISEARCH_ENABLED = getattr(settings, "MEILISEARCH_ENABLED", False)
MEILISEARCH_INDEX_PREFIX = getattr(settings, "MEILISEARCH_INDEX_PREFIX", "")
...
```

What do you think?

pomegranited replied:
Lol.. sounds like a Google situation, where the working approach is deprecated, but the new approach isn't developed yet 😄 

So long as we can pass these settings from the env to the LMS + Studio, I'm good with having them in the `api.py` file.

regisb replied:
OK I'll do that. The only issue I'm facing is that I can't import from api.py because that application is not in the INSTALLED_APPS for the lms -- only for the cms. The api.py module imports models.py , and this fails in the context of the lms. So I'll move the settings to a settings.py module.

pomegranited commented on openedx/core/djangoapps/content/search/engine.py:
```diff
@@ -0,0 +1,484 @@
+        elif isinstance(value, dict):
+            processed[key] = process_document(value)
+        else:
+            # Pray that there are not datetime objects inside lists
```
If [this comment](https://github.com/openedx/edx-platform/pull/35650/files#diff-8146b8d4e4c5363cc67e6830b91ac37a28d6b99a022e49e0a3335a4229fa268cR248) is accurate, then we can just convert datetimes to timestamps like we're doing above, and not worry about their tzoffsets.

regisb replied:
Right. The way we process timezone-aware timestamps is that we convert dict entries from:

```
{
    "mydatetime": datetime(...)
}
```

to:

```
{
    "mydatetime": 1275767, # timestamp
    "mydatetime__tzoffset": 3600 # timezone offset in seconds
}
```

If a datetime entry finds its way in a list, then I no longer know how to convert it in a way that preserves the timezone.

pomegranited replied:
We could at least convert them to UTC timestamps, so at least the deserialized time is accurate, even if the timezone isn't preserved?

regisb replied:
Datetimes in lists will be converted to strings, as per the DocumentEncoder. I don't have other clues to decide whether this is the right decision, as opposed to timestamps...

pomegranited replied:
I guess this is something we can deal with if and when we need to.

pomegranited commented on openedx/core/djangoapps/content/search/engine.py:
```diff
@@ -0,0 +1,484 @@
+            if value.tzinfo:
+                utcoffset = value.utcoffset().seconds
+            processed[key] = value.timestamp()
+            processed[f"{key}{UTC_OFFSET_SUFFIX}"] = utcoffset
```
Hmm.. do we actually need to record the UTC offset? IIRR,

* naive datetimes don't have `tzinfo`, so will record an offset of `None`
* tz-aware datetimes [generate UTC-based timestamps](https://docs.python.org/3/library/datetime.html#datetime.datetime.timestamp), and so the original tzoffset isn't needed in order to understand the timestamp.

regisb replied:
We do need to record the UTC offset to restore the timezone-aware datetime during deserialization. If we convert a timezone-aware datetime to a timestamp, then sure the timestamp will be accurate. But then when we have to convert that timestamp back to a timezone-aware datetime, we won't have enough information.

pomegranited replied:
Hmm ok.. I guess I don't have any experience with use cases where different timezones might be used for different dates throughout the code. But I'm happy to trust your experience :)

pomegranited commented on openedx/core/djangoapps/content/search/engine.py:
```diff
@@ -0,0 +1,484 @@
+
+    We make a copy to avoid modifying the source document.
+    """
+    processed = {PRIMARY_KEY_FIELD_NAME: id2pk(doc["id"])}
```
When I tried reindexing a course, the [recursive call to `process_document`](https://github.com/openedx/edx-platform/pull/35650/files#diff-8146b8d4e4c5363cc67e6830b91ac37a28d6b99a022e49e0a3335a4229fa268cR250) caused an error, because the child `value` dicts didn't contain an `id` field.

```
cms-1  | 2024-10-15 05:20:57,961 ERROR 522 [edx.modulestore] [user 4] [ip 172.20.0.1] courseware_index.py:653 - Course discovery indexing error encountered, course discovery index may be out of date course-v1:OpenedX+DemoX+DemoCourse
cms-1  | Traceback (most recent call last):
cms-1  |   File "/openedx/edx-platform/cms/djangoapps/contentstore/courseware_index.py", line 651, in index_about_information
cms-1  |     searcher.index([course_info])
cms-1  |   File "/openedx/edx-platform/openedx/core/djangoapps/content/search/engine.py", line 120, in index
cms-1  |     [process_document(source) for source in sources],
cms-1  |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
cms-1  |   File "/openedx/edx-platform/openedx/core/djangoapps/content/search/engine.py", line 120, in <listcomp>
cms-1  |     [process_document(source) for source in sources],
cms-1  |      ^^^^^^^^^^^^^^^^^^^^^^^^
cms-1  |   File "/openedx/edx-platform/openedx/core/djangoapps/content/search/engine.py", line 251, in process_document
cms-1  |     processed[key] = process_document(value)
cms-1  |                      ^^^^^^^^^^^^^^^^^^^^^^^
cms-1  |   File "/openedx/edx-platform/openedx/core/djangoapps/content/search/engine.py", line 239, in process_document
cms-1  |     processed = {PRIMARY_KEY_FIELD_NAME: id2pk(doc["id"])}
```

One way to resolve this:

```suggestion
    processed = {}
    if doc.get("id") is not None:
        processed[PRIMARY_KEY_FIELD_NAME] = id2pk(doc["id"])
```

regisb replied:
Good point. I did not re-index documents after I made that change. I'll update the code and the unit test.

pomegranited reviewed (commented):
Thank you for this PR @regisb ! I've tested the course discovery and courseware content search, but need to get forums and notes set up to test these -- will tackle them tomorrow.

Noted a couple of minor issues inline.

regisb commented:
Thanks for the very quick review Jill, I appreciate it!

pomegranited commented on openedx/core/djangoapps/content/search/engine.py:
```diff
@@ -0,0 +1,492 @@
+        "catalog_visibility",  # exclude visibility="none"
+        "enrollment_end",  # include only enrollable courses
+    ],
+    "courseware_content": [
```
Should use [CoursewareSearchIndexer.INDEX_NAME](https://github.com/openedx/edx-platform/blob/e28a01e2bf78d5750c649edf5e8d48776cc38d8d/cms/djangoapps/contentstore/courseware_index.py#L343) instead of a hard-coded string here.

There's surely a variable for `"course_info"` too, but I couldn't find it quickly.

It's unfortunate that we need to store the filterable fields for all the uses of this engine here, instead of near the code that uses them, but I don't see a better way to deal with this..

Should we provide a way for people to inject their own index + fields into this hash?

regisb replied:
> Should use [CoursewareSearchIndexer.INDEX_NAME](https://github.com/openedx/edx-platform/blob/e28a01e2bf78d5750c649edf5e8d48776cc38d8d/cms/djangoapps/contentstore/courseware_index.py#L343) instead of a hard-coded string here.

> There's surely a variable for "course_info" too, but I couldn't find it quickly.

This is actually one of the issues that are raised in this document: https://openedx.atlassian.net/wiki/spaces/AC/pages/3884744738/State+of+edx-search+2023

Basically, index names are hardcoded across multiple repositories in Open edX... which is why I decided ~to be lazy~ to also hardcode the index names. But you're right, let's not aggravate the problem.

I won't go with CoursewareSearchIndexer.INDEX_NAME, because it's in the cms. Instead I suggest to go with COURSEWARE_INFO_INDEX_NAME and COURSEWARE_CONTENT_INDEX_NAME.

What I can do is to change the CoursewareSearchIndexer to pick the index names from the same settings.

> Should we provide a way for people to inject their own index + fields into this hash?

:thinking: yes... but I'd rather avoid adding yet another configuration setting. What I propose is that we leave it up to developers to extend `INDEX_FILTERABLES`. And then in `create_indexes` we concatenate the lists of existing and required filterables, such that we don't accidentally delete any existing index. In addition, we make it possible to pass an `index_filterables` argument to `create_indexes` such that developers can use it to create their own indexes.

pomegranited commented:
Hi @regisb -- my apologies, I ran out of time today to continue testing this for Discussions, but I will pick it up first thing tomorrow.

> 1. Is edx-platform the right place to add this module? (I think it is, because edx-search is used only in edx-platform). If yes, did I pick the right location in the edx-platform file hierarchy, and did I pick the right module name?

@ormsbee @bradenmacdonald What do you think?

regisb committed:
feat: configurable Elasticsearch indices for courseware search

Previously, Elasticsearch indices configured for courseware and course
info storage were hard-coded in the search indexer classes. With this
change, these classes now follow the convention set by edx-search-api
and load the index names from COURSEWARE_CONTENT_INDEX_NAME and
COURSEWARE_INFO_INDEX_NAME. This does not resolve all inconsistencies,
but it's an improvement.

This is not considered a breaking change because when the settings are
not defined they default to the same values: "course_info" and
"courseware_content".

regisb committed:
chore: refactor search settings for reusability

Moving settings from cms/envs/common.py makes it possible to reuse
Meilisearch settings in the LMS without duplicating the settings (see
OEP-45).

regisb commented:
FYI discussions and notes won't work out of the box, because they use their own engines. This is on my TODO-list.

bradenmacdonald commented:
> > 1. Is edx-platform the right place to add this module? (I think it is, because edx-search is used only in edx-platform). If yes, did I pick the right location in the edx-platform file hierarchy, and did I pick the right module name?
> 
> @ormsbee @bradenmacdonald What do you think?

Finding all the code related to edx-search and how it works has always been difficult for me, with the way it's spread out everywhere. I would prefer to locate this closer to some other code that's using edx-search, but I think this is OK. I just want to have a more clear separation between the Studio Search code that doesn't use edx-search APIs at all, and this engine backend that does use edx-search (even though they have a few settings in common, they're totally different implementations and used in totally different ways). So I'd suggest putting the new files in a subfolder, e.g. `openedx.core.djangoapps.content.search.legacy` . This also highlights that we want to replace edx-search with a newer abstraction layer in future versions (or just support meilisearch if we're happy enough with it at that point).

pomegranited commented:
@regisb 

> FYI discussions and notes won't work out of the box, because they use their own engines. This is on my TODO-list.

Ah yes, of course! I realised this when I remembered that we're still using the old cs_comments_service as the default forum :( Good news is there's a [Meilisearch SDK for ruby](https://github.com/meilisearch/meilisearch-ruby), but I still don't envy you or anyone else who has to work on the cs_comments_service code!

@bradenmacdonald 

>  So I'd suggest putting the new files in a subfolder, e.g. openedx.core.djangoapps.content.search.legacy . This also highlights that we want to replace edx-search with a newer abstraction layer in future versions (or just support meilisearch if we're happy enough with it at that point).

I get what you're saying about an intention to replace edx-search, but marking this with "legacy" could be confusing too, since the "legacy" search engine is Elasticsearch. What if we put these new files under the existing [`openedx.features.djangoapps.course_search`](https://github.com/openedx/edx-platform/tree/master/openedx/features/course_search) module?

FYI I ran into weird (unrelated) issues with my tutor dev stack today, so created https://github.com/openedx/edx-platform/pull/35662 to generate a grove k8s sandbox so others can test this out too.

regisb commented:
> Finding all the code related to edx-search and how it works has always been difficult for me, with the way it's spread out everywhere.

I totally get that. In such a case can I suggest moving this code to a separate repo? This would have several advantages:

1. Reusable code in the context of [edx-notes-api](https://github.com/openedx/edx-notes-api/).
2. No need to make changes to edx-platform so close to the release branch cutoff.
3. Easily installable with Tutor.

> Ah yes, of course! I realised this when I remembered that we're still using the old cs_comments_service as the default forum :( Good news is there's a [Meilisearch SDK for ruby](https://github.com/meilisearch/meilisearch-ruby), but I still don't envy you or anyone else who has to work on the cs_comments_service code!

I don't intend to change the ruby cs_comments_service repo. Instead, I'll only work on the new [forum v2 repo](https://github.com/edly-io/forum/) (in the process of being [migrated](https://github.com/openedx/axim-engineering/issues/1272) to the openedx org). This repo will go live in Sumac.

pomegranited commented:
@regisb Separate repo is fine with me.. so why not put this in edx-search ?

> I don't intend to change the ruby cs_comments_service repo. Instead, I'll only work on the new [forum v2 repo](https://github.com/edly-io/forum/) (in the process of being https://github.com/openedx/axim-engineering/issues/1272 to the openedx org). This repo will go live in Sumac.

Ooo spectacular!

regisb commented:
> so why not put this in edx-search ?

That would be the most logical place. @bradenmacdonald you are (understandably) averse to all things related to edx-search, but since this is an implementation of an edx-search interface, do you agree that we should add the code there?

EDIT: here's the PR:  https://github.com/openedx/edx-search/pull/162 I created it early such that we can install edx-search in the Tutor Docker image for Sumac.

bradenmacdonald commented:
@regisb Yes that's totally fine with me! Thanks.

regisb commented:
Closing in favour of https://github.com/openedx/edx-search/pull/162
